### ISSUE TITLE
Performance bottleneck in function maskBytes

### ISSUE THREAD

**Issue by ghjp** [gorilla/websocket#31]
I have investigated the performance of the websocket communication. With the kernel "perf" utility I saw that the function maskBytes consumes a lot of CPU cycles. Therefore I have optimized this function. Now I get twice the data throughout than before (900MB/s instead of 440MB/s). At the moment it is optimized for 64-bit machines. I'm also not sure if the optimization works on big endian machines. The trick is to apply the mask on 64-bit chunks instead of bytes. Here it is:

    import "unsafe"

     func maskBytes(key [4]byte, pos int, b []byte) int {
            // Construct 64 bit mask
            k32 := uint32(key[0]) | uint32(key[1]) << 8 | uint32(key[2]) << 16 | uint32(key[3]) << 24
            k64 := uint64(k32)
            k64 = k64<<32 | k64
            // Correct mask by pos offset
            k64 <<= uint(pos) << 3
            k64 |= uint64(k32 >> ((4 - uint(pos)) << 3))
            // Apply 64 bit mask to the data set
            for len(b) >= 8 {
                    *(*uint64)(unsafe.Pointer(&b[0])) ^= k64
                    b = b[8:]
                    pos += 8
            }
            // Mask the last bytes by a slow method
            for i := range b {
                    b[i] ^= key[pos&3]
                    pos++
            }
            return pos & 3
    }

**Comment by wI2L** [comment]
@garyburd Still interested about this ? I would gladly look into this in details and make.

**Comment by garyburd** [comment]
Yes, I am still interested. 

It's possible to get a big performance gain without writing assembly. The basic idea is to mask a byte a time to a word boundary, use the unsafe package to mask a word at a time, and finish by masking the individual bytes after the last whole word.  See [xor.go](https://github.com/golang/go/blob/master/src/crypto/cipher/xor.go) for a similar approach used in the standard library.  Because the mask can be rotated, it's always possible to use aligned writes.

**Comment by wI2L** [comment]
I benchmarked the current implementation against @ghjp one
```
BenchmarkMaskBytes-8       	 2000000	       656 ns/op	1558.66 MB/s
BenchmarkGhjpMaskBytes-8   	10000000	       122 ns/op	8360.58 MB/s
```

**Comment by garyburd** [comment]
Doh! I should have refreshed my memory on what's in this issue before responding. How does it perform when the the slice is not aligned on a word boundary?

**Comment by wI2L** [comment]
@garyburd I am not very sure of how to simulate this with Go.
Here is the current bench func
```
func BenchmarkMaskBytes(b *testing.B) {
	var key [4]byte
	data := make([]byte, 1024)
	pos := 0
	for i := 0; i < b.N; i++ {
		pos = maskBytes(key, pos, data)
	}
	b.SetBytes(int64(len(data)))
}
```

**Comment by garyburd** [comment]
`make` allocates a word aligned backing array.  Index into the backing array to get an unaligned slice.
 
```
func BenchmarkMaskBytes(b *testing.B) {
	var key [4]byte
	data := make([]byte, 1025)
        data = data[1:]  // <--- backing array of data is not aligned on word boundary.
	pos := 0
	for i := 0; i < b.N; i++ {
		pos = maskBytes(key, pos, data)
	}
	b.SetBytes(int64(len(data)))
}
```

**Comment by wI2L** [comment]
```
BenchmarkMaskBytes-8                	 2000000	       668 ns/op	1530.78 MB/s
BenchmarkFastMaskBytes-8            	20000000	       116 ns/op	8819.79 MB/s
BenchmarkFastMaskBytesUnaligned-8   	 2000000	       665 ns/op	1538.29 MB/s
```

**Comment by wI2L** [comment]
Inspired by the xor example from go sources, I wrote this.

```
const wordSize = int(unsafe.Sizeof(uintptr(0)))

func fastMaskBytes(key [4]byte, pos int, b []byte) int {
	n := len(b)
	w := n / wordSize
	if w > 0 {
		bw := *(*[]uintptr)(unsafe.Pointer(&b))
		for i := 0; i < w; i++ {
			bw[i] ^= uintptr(key[i%3])
		}
	}

	for i := (n - n%wordSize); i < n; i++ {
		b[i] ^= key[i%3]
	}

	return n + pos
}
```

```
BenchmarkMaskBytes-8                	 2000000	       660 ns/op	1549.83 MB/s
BenchmarkMaskBytesUnaligned-8       	 2000000	       671 ns/op	1526.00 MB/s
BenchmarkFastMaskBytes-8            	10000000	       186 ns/op	5500.23 MB/s
BenchmarkFastMaskBytesUnaligned-8   	10000000	       185 ns/op	5505.86 MB/s
```

**Comment by wI2L** [comment]
Tried to write the code with the help of the go xor examples, but I don't know how to mask a whole word using the key, and what i've done does not passe the `TestFraming` test.

**Comment by garyburd** [comment]
Here's a start on the function.

```
const (
	wordSize         = int(unsafe.Sizeof(uintptr(0)))
	simpleMaskCutoff = wordSize * 2
)

func maskBytesSimple(key [4]byte, pos int, b []byte) int {
	for i := range b {
		b[i] ^= key[pos&3]
		pos++
	}
	return pos & 3
}

func maskBytes(key [4]byte, pos int, b []byte) int {
	// TODO: Run benchmarks to determine if simple mask function is faster on
	// small buffers. Adjust simpleMaskCutoff for best performance.
	if len(b) < simpleMaskCutoff {
		return maskBytesSimple(key, pos, b)
	}

	// Mask to aligned position in b.

	if n := int(uintptr(unsafe.Pointer(&b)) % uintptr(wordSize)); n != 0 {
		n = wordSize - n
		pos = maskBytesSimple(key, pos, b[:n])
		b = b[n:]
	}

	// Mask aligned words.

	w := len(b) / wordSize
	if w > 0 { // TODO: Remove this test if simpleMaskCutoff >= 2 * wordSize
		var k [wordSize]byte
		for i := range k {
			k[i] = key[(pos+i)&3]
		}

		bw := *(*[]uintptr)(unsafe.Pointer(&b))
		kw := *(*uintptr)(unsafe.Pointer(&k))
		for i := 0; i < w; i++ {
			// TODO: Use unsafe trickery to avoid bounds check on bw.
			bw[i] ^= kw
		}
	}

	// Mask remaining bytes.

	return maskBytesSimple(key, pos, b[w*wordSize:])
}
```

**Comment by wI2L** [comment]
Tbh, I overvalued my skills for this task, so it seems to me your are basically doing this yourself but I am happy to help a bit for the benchmarks, and at least I am learning from you.
First benchmarks from your implementation looks very promising.
Regarding the unsafe trickery to avoid bounds check i tried this : 
```
			bww := *(*uintptr)(unsafe.Pointer(&bw[i]))
			bww ^= kw
```
but `TestFraming` test failed
```
--- FAIL: TestFraming (0.09s)
	conn_test.go:107: z:false, s:false, r:half, n:65534 c:true: bad byte at offset 1016
	conn_test.go:107: z:false, s:false, r:half, n:65534 c:false: bad byte at offset 1016
	conn_test.go:107: z:false, s:false, r:half, n:65535 c:true: bad byte at offset 1016
	conn_test.go:107: z:false, s:false, r:half, n:65535 c:false: bad byte at offset 1016
	conn_test.go:107: z:false, s:false, r:half, n:65536 c:true: bad byte at offset 1016
	conn_test.go:107: z:false, s:false, r:half, n:65536 c:false: bad byte at offset 1017
	conn_test.go:107: z:false, s:false, r:half, n:65537 c:true: bad byte at offset 1016
	conn_test.go:107: z:false, s:false, r:half, n:65537 c:false: bad byte at offset 1016
	conn_test.go:107: z:false, s:false, r:one, n:124 c:true: bad byte at offset 0
	conn_test.go:107: z:false, s:false, r:one, n:124 c:false: bad byte at offset 0
	conn_test.go:107: z:false, s:false, r:one, n:125 c:true: bad byte at offset 0
	conn_test.go:107: z:false, s:false, r:one, n:125 c:false: bad byte at offset 0
	conn_test.go:107: z:false, s:false, r:one, n:126 c:true: bad byte at offset 0
	conn_test.go:107: z:false, s:false, r:one, n:126 c:false: bad byte at offset 0
	conn_test.go:107: z:false, s:false, r:one, n:127 c:true: bad byte at offset 0
	conn_test.go:107: z:false, s:false, r:one, n:127 c:false: bad byte at offset 0
	conn_test.go:107: z:false, s:false, r:one, n:128 c:true: bad byte at offset 0
	conn_test.go:107: z:false, s:false, r:one, n:128 c:false: bad byte at offset 0
	conn_test.go:107: z:false, s:false, r:one, n:129 c:true: bad byte at offset 0
	conn_test.go:107: z:false, s:false, r:one, n:129 c:false: bad byte at offset 0
	conn_test.go:107: z:false, s:false, r:one, n:65534 c:true: bad byte at offset 0
	conn_test.go:107: z:false, s:false, r:one, n:65534 c:false: bad byte at offset 0
	conn_test.go:107: z:false, s:false, r:one, n:65535 c:true: bad byte at offset 0
	conn_test.go:107: z:false, s:false, r:one, n:65535 c:false: bad byte at offset 0
	conn_test.go:107: z:false, s:false, r:one, n:65536 c:true: bad byte at offset 0
	conn_test.go:107: z:false, s:false, r:one, n:65536 c:false: bad byte at offset 0
	conn_test.go:107: z:false, s:false, r:one, n:65537 c:true: bad byte at offset 0
	conn_test.go:107: z:false, s:false, r:one, n:65537 c:false: bad byte at offset 0
	conn_test.go:107: z:false, s:false, r:asis, n:65534 c:true: bad byte at offset 1016
	conn_test.go:107: z:false, s:false, r:asis, n:65534 c:false: bad byte at offset 1016
	conn_test.go:107: z:false, s:false, r:asis, n:65535 c:true: bad byte at offset 1016
	conn_test.go:107: z:false, s:false, r:asis, n:65535 c:false: bad byte at offset 1016
	conn_test.go:107: z:false, s:false, r:asis, n:65536 c:true: bad byte at offset 1016
	conn_test.go:107: z:false, s:false, r:asis, n:65536 c:false: bad byte at offset 1016
	conn_test.go:107: z:false, s:false, r:asis, n:65537 c:true: bad byte at offset 1016
	conn_test.go:107: z:false, s:false, r:asis, n:65537 c:false: bad byte at offset 1016
	conn_test.go:96: z:true, s:false, r:one, n:124 c:true: ReadFull() returned rbuf, flate: corrupt input before offset 4
	conn_test.go:96: z:true, s:false, r:one, n:124 c:false: ReadFull() returned rbuf, flate: corrupt input before offset 13
	conn_test.go:96: z:true, s:false, r:one, n:125 c:true: ReadFull() returned rbuf, flate: corrupt input before offset 5
	conn_test.go:96: z:true, s:false, r:one, n:125 c:false: ReadFull() returned rbuf, flate: corrupt input before offset 5
	conn_test.go:96: z:true, s:false, r:one, n:126 c:true: ReadFull() returned rbuf, flate: corrupt input before offset 1
	conn_test.go:96: z:true, s:false, r:one, n:126 c:false: ReadFull() returned rbuf, flate: corrupt input before offset 1
	conn_test.go:96: z:true, s:false, r:one, n:127 c:true: ReadFull() returned rbuf, flate: corrupt input before offset 5
	conn_test.go:96: z:true, s:false, r:one, n:127 c:false: ReadFull() returned rbuf, flate: corrupt input before offset 3
	conn_test.go:96: z:true, s:false, r:one, n:128 c:true: ReadFull() returned rbuf, flate: corrupt input before offset 3
	conn_test.go:96: z:true, s:false, r:one, n:128 c:false: ReadFull() returned rbuf, flate: corrupt input before offset 9
	conn_test.go:96: z:true, s:false, r:one, n:129 c:true: ReadFull() returned rbuf, flate: corrupt input before offset 5
	conn_test.go:96: z:true, s:false, r:one, n:129 c:false: ReadFull() returned rbuf, flate: corrupt input before offset 3
	conn_test.go:96: z:true, s:false, r:one, n:65534 c:true: ReadFull() returned rbuf, flate: corrupt input before offset 1
	conn_test.go:96: z:true, s:false, r:one, n:65534 c:false: ReadFull() returned rbuf, flate: corrupt input before offset 9
	conn_test.go:96: z:true, s:false, r:one, n:65535 c:true: ReadFull() returned rbuf, flate: corrupt input before offset 5
	conn_test.go:96: z:true, s:false, r:one, n:65535 c:false: ReadFull() returned rbuf, flate: corrupt input before offset 7
	conn_test.go:96: z:true, s:false, r:one, n:65536 c:true: ReadFull() returned rbuf, flate: corrupt input before offset 15
	conn_test.go:96: z:true, s:false, r:one, n:65536 c:false: ReadFull() returned rbuf, flate: corrupt input before offset 5
	conn_test.go:96: z:true, s:false, r:one, n:65537 c:true: ReadFull() returned rbuf, flate: corrupt input before offset 6
	conn_test.go:96: z:true, s:false, r:one, n:65537 c:false: ReadFull() returned rbuf, flate: corrupt input before offset 4
FAIL
```

**Comment by garyburd** [comment]
See #169.

**Comment by wI2L** [comment]
@garyburd Nice! Perf improvements are really good for 32bytes chunk and upper.
I see you left the TODO mention on the bounds check. Do you think this is what cause the TestFraming to fail ?

**Comment by garyburd** [comment]
I don't know what caused the test to fail. 

I didn't include `*(*uintptr)(unsafe.Pointer(&bw[i]))` because I think the expression does a bounds check.  I need to look at the generated code, but I have not had a chance to do so.

**Comment by garyburd** [comment]
Fixed by 77f110791cfaf7893f72043b54f469352b836d57

Here's the code that I used to avoid the bounds check:

	n := (len(b) / wordSize) * wordSize
	for i := 0; i < n; i += wordSize {
		*(*uintptr)(unsafe.Pointer(uintptr(unsafe.Pointer(&b[0])) + uintptr(i))) ^= kw
	}
